Add WebHook config and implement minigame hooks

And0r- opened the pull request:
A new webhook section has been added to the settings, allowing the configuration of a backend URL and an authentication barrier.
Two webhook options are now available via toggles:

-     Clan Events: Sends a start and stop request for each clan event.
-     Clan Event Series: Sends a start request only for the first event in a series and a stop request only if no new event starts within 2 seconds. This is especially useful for external systems to notify the user when an auto-mode event series is finished and it’s time to return to the PC.

I will test this feature more thoroughly tomorrow, and I welcome any optimization suggestions.

Uraxys commented:
Is there a reason you're using reflections to get the EventType instead of just getting it directly from the minigame instance?

And0r- commented:
I have implemented direct access to EventType without using reflection. Additionally, I've discovered a bug: when a minigame series is finished, the game crashes – apparently, the asynchronous Il2CppSystem.Net.Http call loses the correct execution context after a delay. If you have any ideas or experience with this, I'd appreciate your feedback.

Alternatively, I'm not sure whether it might be more sensible to limit the minigame webhooks to just start/stop and handle series detection externally.

What do you think of the general idea of integrating webhooks into IdlePlus?

Uraxys commented:
If I remember correctly, asynchronous methods don't play nicely when switching between `Il2CppSystem` and `System` types. If you're using `Il2CppSystem.Net.Http`, then use `Il2CppSystem.Threading.Tasks`. I ran into similar crashes early on, especially with `IdlePlus.IdleClansAPI.IdleAPI`. The solution I found was to completely avoid normal `System.Threading` calls and use only the Il2Cpp equivalents.

Regarding the start/stop logic, I agree that keeping webhooks limited to just start/stop makes sense. Tracking series externally should be straightforward, so there's no strong need to add more complexity to the webhook system itself.

I'm all for webhooks, they open up a lot of possibilities for external applications / websites.

And0r- commented:
I've been testing a generic webhook helper.

For example, to send a minigame start webhook, you could call:

```csharp
WebhookHelper.FireWebhookAsync("Minigames", "GET", $"/minigame/start/{minigame.EventType}");
```

This call will:

- Check if "MinigamesEnabled" is true.
- Build the full URL from the base URL in settings and the provided endpoint.
- Add the authorization header if a token is defined.
- Send the GET request to the specified endpoint.

In the settings, I can define an array with the webhook IDs and dynamically generate the toggles.

What do you prefer: should I stick with this unified, generic approach, or would you rather have dedicated methods for each webhook?

Uraxys commented:
I prefer the generic approach. I would also prefer if most of the arguments were replace with an enum or enum like classes, so the `FireWebhookAsync` call only needs the webhook type and the data. The setting ID, request type, and URL path could be stored in the enum itself.  

Just a heads up, I'm planning to refactor the settings system, mainly how settings are registered and how the settings page works. The goal is to support mods/addons that use IdlePlus as a dependency. The same goes for most other systems, it's screaming for a refactor as the project has grown.

And0r- commented:
I've implemented all feedback points and made the PR ready for merging:

Implemented: Generic webhook approach with enums instead of dedicated methods
Direct access to EventType without reflection
Removed series webhook (better implemented on API side)

The current implementation offers:

Configurable URLs and authentication via settings
Robust queue with retry mechanism and exponential backoff
Parallel processing for better performance
Comprehensive metrics and dev commands for testing

Documentation includes detailed inline comments and a WEBHOOKS.md as reference for developers. The system is easily extensible, allowing for additional event types to be added in the future.
I'm open to further feedback, but I believe the implementation is now ready to merge.

Uraxys commented:
Hey! Thank you so much for the PR, and sorry for the delay in getting back to you.

Everything looks great overall! The only thing I'm unsure about is whether we should merge it now or wait for the next game update, as there might be a new feature coming (JWT containing the players username, signed by the game server) that could potentially be used in combination with the authorization header to verify who triggered the webhook. I'd love to hear your thoughts on that.

And0r- commented:
Hello @Uraxys, thank you for your feedback!

I’d merge the PR right away and keep the username in plaintext, since that’s much easier to implement for many small clan projects. As soon as we have the JWT token, we can simply add it to the webhook metadata.

However, if you’d prefer that we drop the plaintext username later and only send the JWT token, we’d need to postpone the merge until then to avoid breaking changes for anyone already relying on the username field.

How would you like to proceed?
